src/net_processing.cpp

```diff
@@ -5820,10 +5830,14 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
                 m_txrequest.ForgetTxHash(gtxid.GetHash());
             }
         }
-
-
         if (!vGetData.empty())
             m_connman.PushMessage(pto, msgMaker.Make(NetMsgType::GETDATA, vGetData));
+
+        // Message: notfound
+        if (!peer->m_notfounds_to_send.empty()) {
+            m_connman.PushMessage(pto, msgMaker.Make(NetMsgType::NOTFOUND, peer->m_notfounds_to_send));
+            peer->m_notfounds_to_send.clear();
+        }
```

Comment:
Can you explain why it is safe to change the order of responses? From a first impression, I think it is possible that a peer sends us a `getdata` [1], then a `ping`, then they would receive the `pong` first, and then the `notfound`, no? Previously they should be in order.

If this is safe to change, it might be good to explain. Also, then it would be better to move this out of the cs_main scope here, because the lock is not needed.

If it is not safe to change, I guess you can just move the send to the process function?

[1] This should happen intermittently whenever there is `fPauseSend` backpressure, or you can force it deterministically by sending an unknown getdata type.

Reply:
- agree that this code can be moved out of `cs_main` 
- I spent some time, and can't think of any specific circumstances that would cause this change in ordering to be problematic in normal node operations (which is not to say its necessarily fine)
- however, one question that came to mind is if this could negatively impact timing in the tests because of the expectations from the `sync_with_ping` function. however, with AJ's patch, any test covering `NOTFOUND` behavior would require some time delay. so this shouldn’t be an issue.